safenode/src/node/api.rs

```diff
@@ -113,118 +112,123 @@ impl Node {
         response_channel: ResponseChannel<Response>,
     ) -> Result<()> {
         trace!("Handling request: {request:?}");
-        match request {
-            Request::Cmd(Cmd::Dbc {
-                signed_spend,
-                source_tx,
-            }) => {
-                self.add_if_valid(signed_spend, source_tx, response_channel)
-                    .await?
-            }
-            Request::Cmd(cmd) => {
-                let resp = self.storage.write(&cmd).await;
-                self.send_response(Response::Cmd(resp), response_channel)
-                    .await;
-            }
-            Request::Query(query) => {
-                let resp = self.storage.read(&query, User::Anyone).await;
-                self.send_response(Response::Query(resp), response_channel)
-                    .await;
-            }
+        let response = match request {
+            Request::Cmd(cmd) => Response::Cmd(self.handle_cmd(cmd).await),
+            Request::Query(query) => Response::Query(self.handle_query(query).await),
             Request::Event(event) => {
```

Comment:
Hey, it looks like this doesn't relate to this PR directly, but I'm just curious, why would an `Event` be handled as a request? We have an event handler task, so just curious why that's not part of this? The doc comments say, "A fact sent to peers." Should it perhaps be called a `Fact`?

Reply:
Event is the proper and and everywhere used terminology.
Events are facts, that's how they are described. Something that happened, past tense.
(Check out event sourcing and CQRS, and you'll see).

Libp2p though has a strange use of cmds, queries and events.
Request/Response is what they (libp2p) call the vehicle for sending msgs between nodes.

Cmds, Queries and Events are the three things that a system - especially a messaging system - sends around. Doesn't matter if it's local or over the wire. The event handler task is for something else, network events. Not domain logic events.

Cmds - imperative; do this! Mutates.
Queries - read only.
Events - facts, it already happened.

So, when something happened at a node, we don't use a cmd to tell other nodes that it shall happen. And we don't ask (query) other nodes about it happening. We tell the world, that it happened.

And therefore we use a request, since that is the bearer of all intra-node messages.

Reply:
Yeah, my first major project back in 2010 or so was a CQRS system with event sourcing. In a way this actually seems to be mixing read and write (both `Cmd` and `Query` is part of `Request`).

In our CQRS system, commands were on the left side and events were on the right, to be handled as a result of a command, which wrote to the event store.

To me, a request is making a demand of something, not a mechanism to make others aware that something happened. Unless you are making a request to ask what happened. Is that what's happening here? For me it just seems strange to see an `Event` as a request. So is the name here being influenced by libp2p?

Reply:
Request is the top level.

A request can be either a Cmd, a Query or an Event. 

It is unfortunate that it is called Request, because it limits it linguistically, when the functionality is that of a _msg_. A message is what it is. We send a msg.

> In our CQRS system, commands were on the left side and events were on the right, to be handled as a result of a command, which wrote to the event store.

Yep, I built a big CQRS system, with event sourcing.

Events were always the results of cmds.
But we don't quite have such a system now. We do direct modification with cmds, and not the functional style where a cmd produces events that are then _applied_ to current state, to reach next state.

Regardless. The problem here is that the naming of Request is unfortunate. We shouldn't limit ourselves because of thjat though. That is the making of libp2p.